### ISSUE TITLE
Cannot parse passwords in some cases

### ISSUE THREAD

**Issue by CelestialGuru** [sergeyklay/django-environ-2#21]
## This passes:
```
> PASSWORD="p^a@s\`s\$w(o%r)d" DATABASE_URL="postgres://user:${PASSWORD}@host:5432/db"; python -c "import environ; print(environ.Env().db('DATABASE_URL'))"
{'NAME': 'db', 'USER': 'user', 'PASSWORD': 'p^a@s`s$w(o%r)d', 'HOST': 'host', 'PORT': 5432, 'ENGINE': 'django.db.backends.postgresql'}
```

### "#" in password
```
> PASSWORD="#" DATABASE_URL="postgres://user:${PASSWORD}@host:5432/db"; python -c "import environ; print(environ.Env().db('DATABASE_URL'))"
{'NAME': '', 'USER': '', 'PASSWORD': '', 'HOST': 'user', 'PORT': '', 'ENGINE': 'django.db.backends.postgresql'}
```
```
> PASSWORD="pass#word" DATABASE_URL="postgres://user:${PASSWORD}@host:5432/db"; python -c "import environ; print(environ.Env().db('DATABASE_URL'))"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.9/site-packages/environ/environ.py", line 268, in db_url
    return self.db_url_config(
  File "/usr/local/lib/python3.9/site-packages/environ/environ.py", line 491, in db_url_config
    'PORT': _cast_int(url.port) or '',
  File "/usr/local/lib/python3.9/urllib/parse.py", line 178, in port
    raise ValueError(message) from None
ValueError: Port could not be cast to integer value as 'pass'
```

**Comment by sergeyklay** [comment]
Hmm.. Looks interesting! I'll try to sort out ASAP.

**Comment by sergeyklay** [comment]
tl;dr: `#` and a few other characters should be URL-encoded/%-encoded when they appear in userinfo which will already parse correctly.

---

Consider the following **wrong** example:

```sh
# .env file contents
DATABASE_URL=postgres://user:#@host:5432/db
```

```py
# settings.py file contents
import environ


env = environ.Env()
result = env.db('DATABASE_URL')

assert result['NAME'] == ''
assert result['USER'] == ''
assert result['PASSWORD'] == ''
assert result['HOST'] == 'user'
assert result['PORT'] == ''
assert result['ENGINE'] == 'django.db.backends.postgresql'
```

Here, the number sign (`#`) was passed as unencoded, raw value.  However, according to [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-2.2), the number sign character must be escaped:


>    URIs include components and subcomponents that are delimited by
>    characters in the "reserved" set.  These characters are called
>    "reserved" because they may (or may not) be defined as delimiters by
>    the generic syntax, by each scheme-specific syntax, or by the
>    implementation-specific syntax of a URI's dereferencing algorithm.
>    If data for a URI component would conflict with a reserved
>    character's purpose as a delimiter, then the conflicting data must be
>    percent-encoded before the URI is formed.
> 
>    reserved    = gen-delims / sub-delims
> 
>    gen-delims  = ":" / "/" / "?" / "#" / "[" / "]" / "@"
> 
>    sub-delims  = "!" / "$" / "&" / "'" / "(" / ")"
>                / "*" / "+" / "," / ";" / "="


Thus, to make this example valid, we have to fix it as follows:

```sh
# .env file contents
DATABASE_URL=postgres://user:%23@host:5432/db
```

```py
# settings.py file contents
import environ


env = environ.Env()
result = env.db('DATABASE_URL')

assert result['NAME'] == 'db'
assert result['USER'] == 'user'
assert result['PASSWORD'] == '#'
assert result['HOST'] == 'host'
assert result['PORT'] == 5432
assert result['ENGINE'] == 'django.db.backends.postgresql'
```

**Comment by sergeyklay** [comment]
The same question: https://github.com/joke2k/django-environ/issues/7#issuecomment-862343054
